src/codegen/cfg.rs

```diff
@@ -324,9 +325,9 @@ impl BasicBlock {
 #[derive(Clone)]
 pub struct ControlFlowGraph {
     pub name: String,
-    pub function_no: Option<usize>,
-    pub params: Vec<Parameter>,
-    pub returns: Vec<Parameter>,
+    pub function_no: FunctionNumber,
+    pub params: Arc<Vec<Parameter>>,
+    pub returns: Arc<Vec<Parameter>>,
```

Comment:
You may avoid the cost of the malloc/memcpy, but are introducing atomic instructions which come with memory barriers for each iteration. I don't know whether this is faster or `clone()` is faster. Up to you

Reply:
Another way of doing this is by keeping a reference to the parameters in the AST, if possible. We do want to add diagnostics in codegen so we can't make the namespace read-only by keeping a reference. Having said that, we would never modify a functions parameters/returns in codegen.

Reply:
I'd rather use a `Rc`, however, the language server is asynchronous, so it requires atomic operations. Keep in mind that the content of an `Arc` is immutable, so the only atomic operations that happen here are incrementing or decrementing the reference counter. Hopefully, they only happen when we call `clone()` and when an `Arc` goes out of scope. There is no race condition from multiple threads reading the elements of an `Arc`. In addition, `Arc` may use atomic operations from the hardware. x86, for instance, [implements some](https://stackoverflow.com/questions/38447226/atomicity-on-x86/38465341#38465341), so I don't expect `Arc` to be slow.

Reply:
> Another way of doing this is by keeping a reference to the parameters in the AST, if possible. We do want to add diagnostics in codegen so we can't make the namespace read-only by keeping a reference. Having said that, we would never modify a functions parameters/returns in codegen.

That's is an option too, but the reference keeper has the same parent as the object the reference points to. I think this is unsafe, due to this answer, you can find on [this thread](https://www.reddit.com/r/rust/comments/dymc8f/selfreference_struct_how_to/):

> Self referential structs are inherently unsafe since they are implicitly invalidated if they are ever moved. This is why the Rust type system doesn't allow this in safe Rust. I will check on the code and get back to you.

Reply:
> I'd rather use a `Rc`, however, the language server is asynchronous, so it requires atomic operations. Keep in mind that the content of an `Arc` is immutable, so the only atomic operations that happen here are incrementing or decrementing the reference counter. Hopefully, they only happen when we call `clone()` and when an `Arc` goes out of scope. There is no race condition arising from multiple threads reading the elements of an `Arc`. In addition, `Arc` may use atomic operations from the hardware. x86, for instance, [implements some](https://stackoverflow.com/questions/38447226/atomicity-on-x86/38465341#38465341), so I don't expect `Arc` to be slow.

atomic operations have to implemented in hardware, else they cannot be atomic. atomic instructions are quite expensive due inter cpu locking protocols.

Reply:
> > Another way of doing this is by keeping a reference to the parameters in the AST, if possible. We do want to add diagnostics in codegen so we can't make the namespace read-only by keeping a reference. Having said that, we would never modify a functions parameters/returns in codegen.
> 
> That's is an option too, but the reference keeper has the same parent as the object the reference points to. I think this is unsafe, due to this answer you can find on [this thread](https://www.reddit.com/r/rust/comments/dymc8f/selfreference_struct_how_to/):

I thought the original is in `Namespace` and the reference would be in `ControlFlowGraph`. That's not the same object.